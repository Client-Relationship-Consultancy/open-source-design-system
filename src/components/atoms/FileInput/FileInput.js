/**
 * FileInput
 */

import React from "react"
import PropTypes from "prop-types"
import styled from "styled-components"
import Button from "../Button"

export const Container = styled.div`
  display: flex;
  align-items: center;
  .inputFile {
    width: 0.1px;
    height: 0.1px;
    opacity: 0;
    overflow: hidden;
    position: absolute;
    z-index: -1;
  }
  button {
    pointer-events: ${props => (props.upload ? "auto" : "none")};
  }
  .inputFile + label {
    cursor: pointer;
  }
`

const ButtonContainer = styled.div`
  display: flex;
  align-items: center;
`

class FileInput extends React.Component {
  state = { selectedFile: null }

  fileChangedHandler = event => this.setState({ selectedFile: event.target.files[0] })

  resetFile = () => this.setState({ selectedFile: null })

  handleUpload = () => this.props.handleUpload(this.state.selectedFile)

  render() {
    return (
      <Container
        upload={this.state.selectedFile}
        id={this.props.id}
        className={this.props.className}
      >
        <input
          type="file"
          onChange={this.fileChangedHandler}
          id="fileInput"
          className="inputFile"
        />
        <ButtonContainer>
          <label htmlFor="fileInput">
            <Button
              onClick={this.handleUpload}
              buttonType={this.props.theme}
              icon={this.state.selectedFile ? "upload" : "file"}
              id="uploadButton"
              type="button"
            >
              {this.state.selectedFile ? this.props.uploadMessage : this.props.chooseFileMessage}
            </Button>
          </label>
          {this.state.selectedFile && (
            <Button buttonType="ghost" icon="times" onClickIcon={this.resetFile} type="button">
              {this.state.selectedFile.name}
            </Button>
          )}
        </ButtonContainer>
      </Container>
    )
  }
}

FileInput.defaultProps = {
  chooseFileMessage: "Choose file",
  uploadMessage: "Upload file",
  theme: "outline",
}

FileInput.propTypes = {
  chooseFileMessage: PropTypes.any,
  handleUpload: PropTypes.func,
  uploadMessage: PropTypes.any,
  theme: PropTypes.string,
  id: PropTypes.string,
  className: PropTypes.string,
}

export default FileInput
