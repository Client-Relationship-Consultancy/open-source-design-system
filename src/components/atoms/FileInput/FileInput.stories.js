import React from "react"
import { storiesOf } from "@storybook/react"

import { text, select } from "@storybook/addon-knobs"

import { FileInput } from "../../../index"

const stories = storiesOf("FileInput", module)

stories.add("Logo or file input", () => {
  const themeDropdown = {
    default: "default",
    ghost: "ghost",
    outline: "outline",
    error: "error",
    complimentary: "complimentary",
  }

  return (
    <FileInput
      handleUpload={() => console.warn("Uploading file...")} // eslint-disable-line no-console
      uploadMessage={text("Upload Message", "Upload")}
      chooseFileMessage={text("Choose File Message", "Choose file...")}
      theme={select("button type", themeDropdown, "outline")}
    />
  )
})
