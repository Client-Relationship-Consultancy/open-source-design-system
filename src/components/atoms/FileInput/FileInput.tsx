/**
 * FileInput
 */

import React from "react";
import styled from "styled-components";
import Button from "../Button";

interface IContainer {
  selectedFile: File | null;
}

export const Container = styled.div<IContainer>`
  display: flex;
  align-items: center;
  .inputFile {
    width: 0.1px;
    height: 0.1px;
    opacity: 0;
    overflow: hidden;
    position: absolute;
    z-index: -1;
  }
  button {
    pointer-events: ${(props) => (props.selectedFile ? "auto" : "none")};
  }
  label {
    cursor: pointer;
  }
`;

Container.displayName = "Container";

const ButtonContainer = styled.div`
  display: flex;
  align-items: center;
`;

ButtonContainer.displayName = "ButtonContainer";

interface IProps extends React.HTMLProps<HTMLInputElement> {
  chooseFileMessage: string;
  uploadMessage: string;
  handleUpload: (file: File) => void;
  theme: string;
}

interface IState {
  selectedFile: File | null;
}

type DefaultProps = Pick<IProps, "chooseFileMessage" | "uploadMessage" | "theme">;

class FileInput extends React.PureComponent<IProps, IState> {
  static defaultProps: DefaultProps = {
    chooseFileMessage: "Choose file",
    uploadMessage: "Upload file",
    theme: "outline",
  };

  state: IState = { selectedFile: null };

  fileChangedHandler = (event: React.ChangeEvent<HTMLInputElement>): void => {
    const file = event.target.files ? event.target.files[0] : null;
    this.setState({ selectedFile: file });
  };

  resetFile = (): void => this.setState({ selectedFile: null });

  handleUpload = (): void => {
    if (this.state.selectedFile !== null) this.props.handleUpload(this.state.selectedFile);
  };

  render(): JSX.Element {
    return (
      <Container
        selectedFile={this.state.selectedFile}
        id={this.props.id}
        className={this.props.className}
      >
        <input
          type="file"
          onChange={this.fileChangedHandler}
          id="fileInput"
          className="inputFile"
        />
        <ButtonContainer>
          <label htmlFor="fileInput">
            <Button
              onClick={this.handleUpload}
              buttonType="primaryOutline"
              icon={this.state.selectedFile ? "upload" : "file"}
              id="uploadButton"
              type="button"
            >
              {this.state.selectedFile ? this.props.uploadMessage : this.props.chooseFileMessage}
            </Button>
          </label>
          {this.state.selectedFile && (
            <Button buttonType="ghost" icon="times" onClick={this.resetFile} type="button">
              {this.state.selectedFile.name}
            </Button>
          )}
        </ButtonContainer>
      </Container>
    );
  }
}

export default FileInput;
