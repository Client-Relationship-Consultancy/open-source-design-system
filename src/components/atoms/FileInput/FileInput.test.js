import React from "react"
import { shallow, mount } from "enzyme"
import "jest-styled-components"
import FileInput, { Container } from "./FileInput"

describe("FileInput test", () => {
  const handleUpload = jest.fn()
  const component = shallow(<FileInput handleUpload={handleUpload} />)
  it("match last snapshot", () => {
    expect(component).toMatchSnapshot()
  })

  it("change the selectedFile in the state", () => {
    expect(component.instance().state).toEqual({ selectedFile: null })
    component.instance().fileChangedHandler({ target: { files: ["hello"] } })
    expect(component.instance().state).toEqual({ selectedFile: "hello" })
  })

  it("match last snapshot with a file selected", () => {
    expect(component).toMatchSnapshot()
  })

  it("calls handleUpload when we click on upload", () => {
    component.instance().fileChangedHandler({ target: { files: ["hello"] } })
    component.find("#uploadButton").simulate("click")
    expect(handleUpload).toHaveBeenCalledWith("hello")
  })

  it("calls resetFile when we click on the cross Icon", () => {
    component.instance().fileChangedHandler({ target: { files: ["hello"] } })
    component.instance().resetFile()
    expect(component.instance().state).toEqual({ selectedFile: null })
  })

  it("doesn't change the selectedFile in the state if event contains no files", () => {
    expect(component.instance().state).toEqual({ selectedFile: null })
    component.instance().fileChangedHandler({ target: {} })
    expect(component.instance().state).toEqual({ selectedFile: null })
  })

  it("uses empty handleUpload function when not given one as a prop", () => {
    const componentWithoutHandleUpload = shallow(<FileInput />)
    const emptyReturn = componentWithoutHandleUpload.instance().handleUpload()
    expect(emptyReturn).toEqual(undefined)
  })
})

describe("Container", () => {
  it("has pointer events auto", () => {
    const container = mount(<Container selectedFile />)
    expect(container).toHaveStyleRule("pointer-events", "auto", {
      modifier: "button",
    })
  })

  it("has pointer events set to none", () => {
    const container = mount(<Container selectedFile={null} />)
    expect(container).toHaveStyleRule("pointer-events", "none", {
      modifier: "button",
    })
  })
})
