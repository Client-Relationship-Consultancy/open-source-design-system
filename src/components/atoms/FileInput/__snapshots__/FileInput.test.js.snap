// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileInput test match last snapshot 1`] = `
<Container
  selectedFile={null}
>
  <input
    className="inputFile"
    id="fileInput"
    onChange={[Function]}
    type="file"
  />
  <ButtonContainer>
    <label
      htmlFor="fileInput"
    >
      <Button
        buttonType="outline"
        icon="file"
        id="uploadButton"
        onClick={[Function]}
        type="button"
      >
        Choose file
      </Button>
    </label>
  </ButtonContainer>
</Container>
`;

exports[`FileInput test match last snapshot with a file selected 1`] = `
<Container
  selectedFile="hello"
>
  <input
    className="inputFile"
    id="fileInput"
    onChange={[Function]}
    type="file"
  />
  <ButtonContainer>
    <label
      htmlFor="fileInput"
    >
      <Button
        buttonType="outline"
        icon="upload"
        id="uploadButton"
        onClick={[Function]}
        type="button"
      >
        Upload file
      </Button>
    </label>
    <Button
      buttonType="ghost"
      icon="times"
      onClickIcon={[Function]}
      type="button"
    />
  </ButtonContainer>
</Container>
`;
